File: Algorithm/BinarySearch/atCoder/B11/B11.py
# 以下は、**Python** を用いた解法です。処理時間とメモリ消費量を考慮し、以下の設計を行っています。

# ---

# ## ✅ アルゴリズム概要

# * **前処理**: 配列 `A` を昇順に **ソート** `O(N log N)`
# * **クエリ処理**: 各 `X` に対して `A[i] < X` を満たす個数を **二分探索**で求める `O(Q log N)`
# * **入力**: `sys.stdin.read()` を使って高速に読み取り
# * **出力**: `sys.stdout.write()` で高速出力（`print()` より効率的）

# ---

# ## ✅ Python実装（型付き）

# ```python
import sys
import bisect
from typing import List

def main() -> None:
    input_data: List[str] = sys.stdin.read().split()
    index: int = 0

    # 配列の長さ N
    N: int = int(input_data[index])
    index += 1

    # 配列 A
    A: List[int] = [int(input_data[i]) for i in range(index, index + N)]
    index += N

    # ソート（前処理） O(N log N)
    A.sort()

    # クエリ数 Q
    Q: int = int(input_data[index])
    index += 1

    # 各クエリ Xi に対して A 中の「Xi 未満の個数」を求める O(Q log N)
    results: List[int] = []
    for i in range(Q):
        X: int = int(input_data[index + i])
        count: int = bisect.bisect_left(A, X)  # A[i] < X の個数
        results.append(count)

    # 出力（高速） O(Q)
    sys.stdout.write('\n'.join(map(str, results)) + '\n')

if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 処理時間・メモリの見積もり

# ### 時間計算量

# | 処理内容      | 時間計算量                         |
# | --------- | ----------------------------- |
# | 入力の読み込み   | O(N + Q)                      |
# | 配列 A のソート | O(N log N)                    |
# | 各クエリの処理   | O(Q log N)                    |
# | 合計        | O(N log N + Q log N)（≒ 2×10⁶） |

# ✔️ 最大 `N = Q = 10^5` のときでも **余裕で 1秒以内**

# ---

# ### メモリ使用量（目安）

# | 要素      | サイズ（概算）                   |
# | ------- | ------------------------- |
# | 配列 `A`  | 約 `100,000 × 4B` = 400 KB |
# | クエリ `Q` | 約 `100,000 × 4B` = 400 KB |
# | その他     | 数MB以下                     |

# ✔️ 全体で **数MB** 程度 → 1024 MiB 制限に **余裕で収まる**

# ---

# ## ✅ 入力例の確認

# ### 入力

# ```
# 5
# 1 3 3 3 1
# 2
# 4
# 3
# ```

# ### 出力

# ```
# 5
# 2
# ```

# ---

# ご希望あれば、`matplotlib` を使って探索の過程を図解することも可能です。必要であればお知らせください。

File: Algorithm/BinarySearch/atCoder/B11/B11.ts
// 以下は、**TypeScript** + `fs` モジュールを用いて高速に入力を処理し、各クエリ `X` に対して「配列 A の中で X より小さい要素の個数」を求めるプログラムです。

// ---

// ### ✅ アルゴリズム概要（再掲）

// * 配列 `A` を昇順に **ソート**（`O(N log N)`）
// * 各クエリ `X` に対して **二分探索**で `X` 未満の要素数を求める（`O(log N)` × Q）

// ---

// ### ✅ TypeScript 実装（Node.js + fs）

// ```ts
import * as fs from 'fs';

// 入力読み込み（高速）
const input: string[] = fs.readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/);
let idx = 0;

// 入力取得
const N: number = parseInt(input[idx++]);
const A: number[] = [];

for (let i = 0; i < N; i++) {
    A.push(parseInt(input[idx++]));
}

A.sort((a, b) => a - b); // 昇順ソート

const Q: number = parseInt(input[idx++]);
const results: number[] = [];

// 各クエリに対して lowerBound を実行
for (let i = 0; i < Q; i++) {
    const X: number = parseInt(input[idx++]);
    results.push(lowerBound(A, X));
}

// 二分探索：X より小さい要素の数を返す
function lowerBound(arr: number[], target: number): number {
    let left = 0;
    let right = arr.length;
    while (left < right) {
        const mid = (left + right) >>> 1; // unsigned right shift で高速化
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

// 結果出力
console.log(results.join('\n'));
// ```

// ---

// ### ✅ 処理時間・メモリの分析

// * **時間計算量**:

//   * `sort`：`O(N log N)`
//   * `Q` 個のクエリに対して `O(log N)` → 合計 `O(Q log N)`
//   * 制約最大 `N=10^5`, `Q=10^5` でも合計約 `2×10^6` 回の操作 → **十分高速**（1秒以内）

// * **メモリ使用量**:

//   * `A`, `results`, 入力配列を保持 → 最大で `300,000` 要素程度
//   * 各要素は整数（8バイト以下） → **数MBで収まる**（1024 MiB 制限内）

// ---

// ### ✅ 入力例での確認

// 入力:

// ```
// 15
// 83 31 11 17 32 19 23 37 43 47 53 61 67 5 55
// 5
// 10
// 20
// 30
// 40
// 50
// ```

// 出力:

// ```
// 1
// 4
// 5
// 8
// 10
// ```

// ---

// ご希望あれば、`readline` を使った逐次読み込み版や、ブラウザ実行向けへの変換、型の強化（ジェネリクス）なども可能です。お気軽にどうぞ。
